### PR TITLE
feat: cal ai

### PR DESCRIPTION
**Pull request by DexterStorey**
This PR is a draft, we plan on moving it out of draft by EOD, but are still fixing up issues with app installation.
There also seems to be an issue with the core getEventTypes API in main, where it works in prod but not locally.

## What does this PR do?

Demo: https://www.loom.com/share/950b0f03e1534cad92295e4720440ab4?sid=0eb02880-ea97-4d6a-9586-9582d8eb9caf

This pr adds the Cal AI app to the cal apps folder and the App Store.

The Cal AI app is an email chat interface that allows you to:
- Chat with your calendar: "What meetings do I have today?"
- Book Meetings: "Book a meeting with bob@example.com on Friday at 2pm"
- Update Meetings: "Push the meeting with bob an hour back"
- Delete Meetings: "Cancel the meeting with Bob"
- More complicated commands (subject to vercel function timeout): "Cancel all my meetings this week, push anything with Bob to next week and book a meeting with bob called 'Take me to the hospital' right now"

The Cal AI app is a new application in the /apps folder that should be deployed to https://cal.ai

## Requirement/Documentation

- If there is a requirement document, please, share it here.
- If there is ab UI/UX design document, please, share it here.

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## How should this be tested?

- ~~Are there environment variables that should be set?~~
- ~~What are the minimal test data to have?~~
- ~~What is expected (happy path) to have (input and output)?~~
- ~~Any other important info that could help to test that PR~~

You will need:

- A running instance of the cal.com monorepo with a db
- An admin access token
- An Open AI account with access to GPT 4
- A Resend Account
- A Sendgrid Account with inbound parse configured

- Populate the .env file in apps/ai based on the .env.example file 

- Run the web, api and ai apps `yarn dev:ai`

- Navigate to the web app and install the cal ai app

- Email the email you configured with inbound parse any instructions

## Performance

Email has inherent speed limitations based on your client.
Apple writes in real time but reads every minute
Gmail reads in real time but writes 10 seconds later because of mandatory “undo send” feature
Additionally, Resend takes some time to send

The AI agent loop can be benchmarked with [langsmith](https://smith.langchain.com)
For example, check out [this run](https://smith.langchain.com/public/67d18947-48c7-47a0-ac95-0f2c45078e03/r)

Screenshot 2023-08-26 at 1 44 24 AM

On complex tasks with many steps, vercel can timeout the agent if it goes longer than 1 minute on a pro plan.

## Mandatory Tasks

- [x] Make sure you have self-reviewed the code. A decent size PR without self-review might be rejected.

## Checklist

- ~~I haven't read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)~~
- ~~My code doesn't follow the style guidelines of this project~~
- ~~I haven't commented my code, particularly in hard-to-understand areas~~
- ~~I haven't checked if my PR needs changes to the documentation~~
- ~~I haven't checked if my changes generate no new warnings~~
- ~~I haven't added tests that prove my fix is effective or that my feature works~~
- ~~I haven't checked if new and existing unit tests pass locally with my changes~~

### CODE REVIEW

File: apps/ai/src/tools/getEventTypes.ts
Comment (DexterStorey) [inline apps/ai/src/tools/getEventTypes.ts:19]: This should be removed. The getEventTypes call works in production, but it seems like there is an issue with getEventTypes in the main branch in the api when hosting locally. Once this issue is fixed, we can remove this temporary code and it should work.

File: packages/app-store/cal-ai/api/add.ts
Comment (DexterStorey) [inline packages/app-store/cal-ai/api/add.ts:1]: The app installs successfully, but the UI still shows the install button.
Working on a fix.

**Comment by DexterStorey** [comment]
Created this branch: https://github.com/RubricLab/cal.com/tree/dexterstorey/event-types-fix
Which is a quick fix for https://github.com/calcom/cal.com/issues/10995

This allows calls to localhost:3002/api/event-types?apiKey=API_KEY to work, so that you can run this locally.

**Comment by DexterStorey** [comment]
> Created this branch: https://github.com/RubricLab/cal.com/tree/dexterstorey/event-types-fix Which is a quick fix for #10995
> 
> This allows calls to localhost:3002/api/event-types?apiKey=API_KEY to work, so that you can run the ai app locally.

This is now fixed! Commented on the issue.

File: apps/ai/src/tools/getEventTypes.ts
Comment (DexterStorey) [reply apps/ai/src/tools/getEventTypes.ts]: Removed!

**Comment by hariombalhara** [comment]
Wow!! So excited to review this !!!!

File: apps/ai/.env.example
Comment (PeerRich) [inline apps/ai/.env.example:15]: @hariombalhara i think this should be in the appStore.env no?

File: apps/ai/.env.example
Comment (PeerRich) [inline apps/ai/.env.example:9]: i wonder, should we use sendgrid here?

File: apps/ai/README.md
Comment (PeerRich) [inline apps/ai/README.md:28]: sendgrid maybe?

File: apps/ai/.env.example
Comment (nicktrn) [reply apps/ai/.env.example:9]: Any particular reason to not use nodemailer with existing env vars like `EMAIL_FROM`?

Comment (tedspare) [reply apps/ai/.env.example:9]: Migrating to nodemailer.

As for `EMAIL_FROM`, the only concern would be replying `notifications@cal.com` (or the actual value) after the user emailed `hi@cal.ai`.

Come to think of it, this should be dynamic, allowing emails to `*@cal.ai`.

File: apps/ai/README.md
Comment (tedspare) [reply apps/ai/README.md]: Done

**Comment by DexterStorey** [comment]
This is ready for review! 🦾🎉

**Comment by DexterStorey** [comment]
@PeerRich can you authorize the deployment?

**Comment by PeerRich** [comment]
fantastic, cant wait to use this!

File: apps/ai/.env.example
Comment (nicktrn) [inline apps/ai/.env.example:1]: This seems unused now

File: apps/ai/src/app/api/receive/route.ts
Comment (nicktrn) [inline apps/ai/src/app/api/receive/route.ts:31]: Can this be validated? Sendgrid offers an example payload of `"{@sendgrid.com : pass}"` - will this be empty on failure or maybe set to `fail`?

File: apps/ai/src/app/api/receive/route.ts
Comment (nicktrn) [inline apps/ai/src/app/api/receive/route.ts:58]: See `signature` comment above. Should be fine if this is expected to be falsy on DKIM verification failure.

File: packages/app-store/cal-ai/config.json
Comment (nicktrn) [inline packages/app-store/cal-ai/config.json:12]: Email does not match package.json description

File: packages/app-store/cal-ai/package.json
Comment (nicktrn) [inline packages/app-store/cal-ai/package.json:13]: Email does not match config.json description

File: apps/ai/src/utils/verifyParseKey.ts
Comment (nicktrn) [inline apps/ai/src/utils/verifyParseKey.ts:7]:
```suggestion
 * env.PARSE_KEY must be configured as a query param in the sendgrid inbound parse settings.
```

File: apps/ai/.env.example
Comment (DexterStorey) [reply apps/ai/.env.example]: Fixed 🫡

File: packages/app-store/cal-ai/package.json
Comment (DexterStorey) [reply packages/app-store/cal-ai/package.json]: [Opted for hi@cal.ai](https://github.com/calcom/cal.com/pull/10992/commits/7eebfa1e8e12ec61b20f6fb59c7b03bf45310a8f) Thanks for pointing this out! 🙌

File: packages/app-store/cal-ai/config.json
Comment (DexterStorey) [reply packages/app-store/cal-ai/config.json:12]: updated package.json!

File: apps/ai/src/app/api/receive/route.ts
Comment (DexterStorey) [reply apps/ai/src/app/api/receive/route.ts:31]: Great catch - Will red team it with a non verified email and see what the payload is.

Comment (nicktrn) [reply apps/ai/src/app/api/receive/route.ts:31]: Would hope that SendGrid doesn't forward emails that fail DKIM. But then the question is, what happens if DKIM is not configured? For security, DKIM setup and verification should be a hard requirement.

Comment (DexterStorey) [reply apps/ai/src/app/api/receive/route.ts]: Agreed.

See [dca8cd5](https://github.com/calcom/cal.com/pull/10992/commits/dca8cd54492b432ce15c9b7734b0fd9e0167ab54)

File: apps/ai/src/app/api/receive/route.ts
Comment (DexterStorey) [reply apps/ai/src/app/api/receive/route.ts:59]: signature should now be false if DKIM fails or isn't passed! 🤙

**Review by PeerRich — DISMISSED** [review]
approving the flow, waiting for @emrysal for ultimate approval of code.

**Review by emrysal — DISMISSED** [review]
Love to see this on dev 🙏

File: apps/ai/src/app/api/agent/route.ts
Comment (zomars) [inline apps/ai/src/app/api/agent/route.ts:30]: Wouldn't this cause context mixups if a lambda server multiple requests per invocation? 

I've seen this happen before. When we have multiple requests per second, there's the chance that a previous context gets overwritten by a newer request and request A gets data from request B and so on.

Comment (zomars) [reply apps/ai/src/app/api/agent/route.ts:30]: Some more context: https://adamtuttle.codes/blog/2022/aws-lambda-safe-global-variables/

File: apps/ai/src/utils/context.ts
Comment (zomars) [inline apps/ai/src/utils/context.ts:1]: Same as my comment above.

File: apps/ai/src/utils/now.ts
Comment (zomars) [inline apps/ai/src/utils/now.ts:1]: clever!

**Review by zomars — DISMISSED** [review]
Awesome work so far @DexterStorey. Just have a possible major concern on global variables. Blocking until that is addressed. Other than that the code looks awesome. 🙏🏽

File: apps/ai/src/app/api/agent/route.ts
Comment (DexterStorey) [reply apps/ai/src/app/api/agent/route.ts:30]: thanks so much @zomars for this awesome review and for finding this. Working on it now.

**Comment by DexterStorey** [comment]
Battle tested Ted's global vars fix with the following runs. 
Looks like all the tools are working.

@zomars can you take a look at [5450692](https://github.com/calcom/cal.com/pull/10992/commits/545069216227c1a89b839adb50ef65777d755c9f)?

[Book a meeting at a good time](https://smith.langchain.com/public/becc0ee2-f8f8-4848-b51d-70a898e7fbf5/r)
[Cancel all my meetings](https://smith.langchain.com/public/0860e766-a2c2-43b3-b3ca-03d0e2cad758/r)
[Move all my meetings](https://smith.langchain.com/public/3e9ea1e9-f95e-4f18-b7a9-f7305002aa7f/r)

**Review by PeerRich — APPROVED** [review]
re-requesting @zomars approval

**Review by zomars — APPROVED** [review]
Nice! Thank you for your contribution 🙏 

Ship it

**Pull request merged**
